### PR TITLE
feat(UI): Implement a `DraggableList` component

### LINKED ISSUE #408: Add capabilities to drag'n'drop card in the layout
# Feature Request

_User cant reorganize easily there cards in the view builder._

One solution is to add a handle to each cards and implement a drag'n'drop system to reorder cards.
One alternative coudl be to use a dedicated grid layout library like [grid layout plus](https://grid-layout-plus.netlify.app/)

> Here is a link to the drag&drop flow @rouk1  : https://www.figma.com/design/Qzjyxv0lPDU2kKtsH9dcp3/Mandr---Workstation---High-fidelity-mock-ups---Q3?node-id=3632-3791&t=r3d92IJPyjWvDeVI-4 

 _Originally posted by @anasstarfaoui in [#326](https://github.com/probabl-ai/skore/issues/326#issuecomment-2382379271)_

### PR DESCRIPTION
**Pull request by rouk1**
This PR is the first step to fix #408.

## UI/UX preview

https://github.com/user-attachments/assets/d9db0c12-bee1-4795-ab3b-fc7c4e11ae9e

## usage

The component needs a reactive list of `Item` (a ref). Item is a regular object that needs an id filed that is a string.
It will mutate the given array of items to reflect the user selected order.

```vue
<script setup lang="ts">
import { ref } from "vue";

const draggableListData = ref(
  Array.from({ length: 25 }, (v, i) => ({
    id: `${i}`,
    color: `hsl(${(360 / 25) * i}deg, 90%, 50%)`,
    content: Array.from(
      { length: Math.floor(Math.random() * 10) + 1 }, // Random number of items between 1 and 10
      () => String.fromCharCode(97 + Math.floor(Math.random() * 26)) // Random lowercase letter
    ),
  }))
);
</script>

<template>
  <DraggableList v-model:items="draggableListData">
    <template #item="{ id, color, content }">
      <div :style="{ backgroundColor: color, color: 'white' }">
        <span>ID: {{ id }}</span>
        <ul>
          <li v-for="(c, i) in content" :key="i">{{ c }}</li>
        </ul>
      </div>
    </template>
  </DraggableList>
</template>
```

## implementation choices

The implementation introduces two new dependencies. 
- [html-to-image](https://github.com/bubkoo/html-to-image) to create rasterized version of the moved item
- [interact.js](https://interactjs.io/) to streamline pointer interaction

The main container is a flexbox, items flow normally in it (no absolute positioning).
When a drag start the selected items becomes slightly transparent and a raster copy of it is generated. That raster copy follows user's pointer using computed translation (CSS transform). To make room for the item to be dropped each items also has a drop indicator. The size of this drop indicator reflects the size of the dragged item. I use margin top & bottom to give different feedback when dragging down or up. This is nice because no coordinates calculation is needed and items continue to flow normally in the container.

To limit clashes with inner item code when dragging all the items are transformed to a rasterized version of themself.

## known limitation

~~For now it is not possible to automatically scroll the container when dragging outside of its visible range.~~

### CODE REVIEW

**Comment by tuscland** [comment]
In the video, there's a moment where you are unable to drag the item. It looks like a bug, but I can't be sure.

**Comment by rouk1** [comment]
Edit:  update PR descritption (enable autoscroll & raster cache)

File: skore-ui/src/components/CacheableComponent.vue [hunk 1]
```diff
@@ -0,0 +1,77 @@
+<script setup lang="ts">
```
Comment (tuscland) [inline skore-ui/src/components/CacheableComponent.vue:1]: In some way "cacheable" means that the component uses resources to build or use, and in order to spare those resources, we cache it.

So I was thinking a better name could be "DynamicContentRasterizer".
And here is the source code for it:

```vue
<script setup lang="ts">
import { debounce } from "@/services/utils";
import { toPng } from "html-to-image";
import { computed, onBeforeUnmount, ref, useTemplateRef, watch } from "vue";

const props = defineProps<{
  isRasterized: boolean;
}>();

const rasterizedSrc = ref("");
const realContent = useTemplateRef("realContent");
const contentHeight = ref(0);

const styles = computed(() => ({
  height: contentHeight.value !== 0 ? `${contentHeight.value}px` : "auto",
}));

const updateRasterizedImage = debounce(
  async () => {
    if (realContent.value) {
      rasterizedSrc.value = await toPng(realContent.value as HTMLElement);
    }
  },
  100,
  false
);

const resizeObserver = new ResizeObserver(async (entries) => {
  if (entries.length == 1 && !props.isRasterized) {
    const content = entries[0];
    contentHeight.value = content.contentRect.height;
    updateRasterizedImage();
  }
});

watch(
  () => realContent.value,
  () => {
    if (realContent.value) {
      resizeObserver.observe(realContent.value as HTMLElement);
    }
  }
);

onBeforeUnmount(() => {
  resizeObserver.disconnect();
});
</script>

<template>
  <div class="dynamic-content-rasterizer" :style="styles">
    <div class="real-content" v-if="!isRasterized" ref="realContent">
      <slot />
    </div>
    <div class="rasterized-content" v-else>
      <img :src="rasterizedSrc" alt="Rasterized content" />
    </div>
  </div>
</template>

<style scoped>
.dynamic-content-rasterizer {
  position: relative;
  padding: 0;
  margin: 0;
}

.rasterized-content {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
}
</style>
```

File: skore-ui/src/components/DraggableList.vue [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+    modifiers: [makeModifier()],
+    listeners: {
+      async start(event) {
+        // make a rasterized copy of the moving element
+        const content = event.target.parentElement.querySelector(".content");
+        if (content && container.value) {
+          const parentBounds = container.value.getBoundingClientRect();
+          const bounds = content.getBoundingClientRect();
+
+          movingItemAsPngData.value = await toPng(content);
```
Comment (tuscland) [inline skore-ui/src/components/DraggableList.vue:120]: Sorry, I surely miss something here. Why is it needed to raster the component manually? Isn't it the purpose of the specialized `CacheableComponent`?

**Review by tuscland — COMMENTED** [review]
Only a couple of suggestions.

File: skore-ui/src/components/DraggableList.vue (hunk 2 above)
Comment (rouk1) [reply skore-ui/src/components/DraggableList.vue:120]: This one is the one that follows the cursor.

File: skore-ui/src/components/CacheableComponent.vue (hunk 1 above)
Comment (rouk1) [reply skore-ui/src/components/CacheableComponent.vue]: Rename done : )

**Review by thomass-dev — APPROVED** [review]

**Pull request merged**